### PR TITLE
feat: optimistic updates and defence for follow/unfollow

### PR DESCRIPTION
**Pull request by reecejohnson**
- Adds optimistic update of UI for follow
- Adds defence in `useUnfollow` hook is a profile is only optimistically following another and attempts to unfollow

### CODE REVIEW

File: packages/react/src/transactions/useUnfollow.ts [hunk 1]
```diff
@@ -30,6 +35,15 @@ export function useUnfollow({ profile }: UseUnfollowArgs) {
...
   return {
     unfollow: async () => {
+      if (!profile.isFollowedByMe && profile.isOptimisticFollowedByMe) {
+        setError(
+          new OptimisticUnfollowError(
+            `Your follow request for ${profile.handle} is still pending.`,
+          ),
+        );
+        return;
+      }
```
Comment (cesarenaldi) [inline packages/react/src/transactions/useUnfollow.ts:45]: Neat!

File: packages/domain/src/use-cases/profile/UnfollowProfile.ts [hunk 2]
```diff
@@ -10,6 +10,12 @@ export type UnfollowRequest = {
   kind: TransactionKind.UNFOLLOW_PROFILE;
 };
 
+export class OptimisticUnfollowError extends Error {
+  constructor(message: string) {
+    super(message);
+  }
+}
```
[the remark this answers is not in the document]
Comment (reecejohnson) [reply packages/domain/src/use-cases/profile/UnfollowProfile.ts:17]: okay yeah, makes sense.

i like `PrematureUnfollow`

File: packages/react/src/transactions/useFollow.ts [hunk 3]
```diff
@@ -97,35 +92,36 @@ export function useFollow({ profile }: UseFollowArgs) {
     | WalletConnectionError
     | null
   >(null);
+  const [isPending, setIsPending] = useState<boolean>(false);
   const follow = useFollowController();
   const activeWallet = useActiveWallet();
   const { data: activeProfile } = useActiveProfile();
```
Comment (desfero) [inline packages/react/src/transactions/useFollow.ts:98]: interesting observation that `useActiveWallet` and `useActiveProfile` do not follow the same approach of returning `wallet/profile`

Comment (reecejohnson) [reply packages/react/src/transactions/useFollow.ts:98]: It is interesting, both are using reactive vars but active profile has this logic:

```typescript
if (state === ApplicationsState.LOADING) {
    return {
      loading: true,
      data: undefined,
    };
  }
  ```
  
  do we know the reason for the discrepancy?  @cesarenaldi @desfero

**Review by krzysu — APPROVED** [review]
👍

File: packages/react/src/transactions/useFollow.ts (hunk 3 above)
Comment (desfero) [reply packages/react/src/transactions/useFollow.ts:98]: it appears as a precaution to not start loading queries until we know that the app signaled readiness.
I can see that when the code evolves here (https://github.com/lens-protocol/lens-sdk/blob/3556929122e89389bf9d4640afd03810a0516065/packages/domain/src/use-cases/lifecycle/Bootstrap.ts#L70-L77) we can end up with some race condition between a call to `activeProfile.loadActiveProfileByOwnerAddress` and `applicationPresenter.signalReady.

Comment (reecejohnson) [reply packages/react/src/transactions/useFollow.ts:98]: after line 73 in the code above, we know for sure that we have a an active profile, so it looks like the checks are fine?

Comment (desfero) [reply packages/react/src/transactions/useFollow.ts:98]: i guess soo, maybe @cesarenaldi knows some edge cases?

Comment (cesarenaldi) [reply packages/react/src/transactions/useFollow.ts:98]: It has to do with avoiding flash of logged-out content. To know if we have done-done and if we have a profile or not the only reliable way was to wait for the ready signal.

Arguably we can shift the logged-in/out responsibility to the `useActiveWallet` and let the `useActiveProfile` just propagate the content of the Reactive Var, thoughts?

Comment (reecejohnson) [reply packages/react/src/transactions/useFollow.ts:98]: Continuing this conversation here: https://avara.height.app/T-3548 🫡

**Pull request merged**